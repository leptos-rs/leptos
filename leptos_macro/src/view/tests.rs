use proc_macro2::TokenStream;
use std::str::FromStr;
use syn::parse_quote;

fn pretty(input: TokenStream) -> String {
    let type_item: syn::Item = parse_quote! {
        fn view(){
            #input
        }
    };

    let file = syn::File {
        shebang: None,
        attrs: vec![],
        items: vec![type_item],
    };

    prettyplease::unparse(&file)
}

macro_rules! assert_snapshot
{
    (@assert text $result:ident) => {
        insta::assert_snapshot!(pretty($result))
    };
    (@assert full $result:ident) => {
        insta::assert_debug_snapshot!($result)
    };
    (client_template($assert:ident) => $input: expr) => {
        {
            let tokens = TokenStream::from_str($input).unwrap();
            let nodes = rstml::parse2(tokens).unwrap();
            let result = crate::view::client_template::render_template(&&nodes);

            assert_snapshot!(@assert $assert result)
        }
    };
    (client_builder($assert:ident) => $input: expr) => {
        {
            let tokens = TokenStream::from_str($input).unwrap();
            let nodes = rstml::parse2(tokens).unwrap();
            let mode = crate::view::Mode::Client;
            let global_class = None;
            let call_site = None;
            let result = crate::view::render_view(&&nodes, mode, global_class, call_site);

            assert_snapshot!(@assert $assert result)
        }
    };
    (server_template($assert:ident) => $input: expr) => {
        {
            let tokens = TokenStream::from_str($input).unwrap();
            let nodes = rstml::parse2(tokens).unwrap();
            let mode = crate::view::Mode::Ssr;
            let global_class = None;
            let call_site = None;
            let result = crate::view::render_view(&&nodes, mode, global_class, call_site);

            assert_snapshot!(@assert $assert result)
        }
    }

}

macro_rules! for_all_modes {
    (@ $module: ident, $type: ident => $(
        $test_name:ident => $raw_str:expr
    ),*
    ) => {
        mod $module {
            use super::*;
            $(
                #[test]
                fn $test_name() {
                    assert_snapshot!($type(text) => $raw_str)
                }
            )*
            mod full_span {
                use super::*;
                $(
                    #[test]
                    fn $test_name() {
                        assert_snapshot!($type(full) => $raw_str)
                    }
                )*
            }

        }
    };
    (   $(
        $tts:tt
        )*
    ) => {
        for_all_modes!{@ csr, client_builder =>  $($tts)*}
        for_all_modes!{@ client_template, client_template => $($tts)*}
        for_all_modes!{@ ssr, server_template => $($tts)*}
    };

}

for_all_modes! {
    test_simple_counter => r#"
        <div>
            <button on:click=move |_| set_value(0)>"Clear"</button>
            <button on:click=move |_| set_value.update(|value| *value -= step)>"-1"</button>
            <span>"Value: " {value} "!"</span>
            <button on:click=move |_| set_value.update(|value| *value += step)>"+1"</button>
        </div>
    "#,
    test_counter_component => r#"
        <SimpleCounter
            initial_value=0
            step=1
        />
    "#,
    test_custom_event => r#"
        <ExternalComponent on:custom.event.clear=move |_: Event| set_value(0) />
    "#
}
