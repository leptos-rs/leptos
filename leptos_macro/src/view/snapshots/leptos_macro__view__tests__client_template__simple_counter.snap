---
source: leptos_macro/src/view/tests.rs
expression: pretty(result)
---
fn view() {
    {
        thread_local! {
            static __TEMPLATE : ::leptos::web_sys::HtmlTemplateElement = { let document =
            ::leptos::document(); let el = document.create_element("template").unwrap();
            el
            .set_inner_html("<div><button>Clear</button><button>-1</button><span>Value: <!>!</span><button>+1</button></div>");
            ::leptos::wasm_bindgen::JsCast::unchecked_into(el) }
        }
        let _ = ::leptos::leptos_dom::html::div;
        let _ = ::leptos::leptos_dom::html::div;
        let _ = ::leptos::leptos_dom::html::button;
        let _ = ::leptos::leptos_dom::html::button;
        let _ = ::leptos::leptos_dom::html::button;
        let _ = ::leptos::leptos_dom::html::button;
        let _ = ::leptos::leptos_dom::html::span;
        let _ = ::leptos::leptos_dom::html::span;
        let _ = ::leptos::leptos_dom::html::button;
        let _ = ::leptos::leptos_dom::html::button;
        let root = __TEMPLATE
            .with(|tpl| tpl.content().clone_node_with_deep(true))
            .unwrap()
            .first_child()
            .unwrap();
        let _el1 = "div";
        let _el1 = ::leptos::wasm_bindgen::JsCast::unchecked_into::<
            leptos::web_sys::Node,
        >(root.clone());
        let _el2 = "button";
        let _el2 = _el1
            .first_child()
            .unwrap_or_else(|| panic!("error: {} => {}", "button", "firstChild"));
        let _el3 = _el2
            .first_child()
            .unwrap_or_else(|| panic!("error : {} => {} ", "{block}", "firstChild"));
        let _el4 = "button";
        let _el4 = _el2
            .next_sibling()
            .unwrap_or_else(|| panic!("error : {} => {} ", "button", "nextSibling"));
        let _el5 = _el4
            .first_child()
            .unwrap_or_else(|| panic!("error : {} => {} ", "{block}", "firstChild"));
        let _el6 = "span";
        let _el6 = _el4
            .next_sibling()
            .unwrap_or_else(|| panic!("error : {} => {} ", "span", "nextSibling"));
        let _el7 = _el6
            .first_child()
            .unwrap_or_else(|| panic!("error : {} => {} ", "{block}", "firstChild"));
        let _el8 = _el7
            .next_sibling()
            .unwrap_or_else(|| panic!("error : {} => {} ", "{block}", "nextSibling"));
        let _el9 = _el8
            .next_sibling()
            .unwrap_or_else(|| panic!("error : {} => {} ", "{block}", "nextSibling"));
        let _el10 = "button";
        let _el10 = _el6
            .next_sibling()
            .unwrap_or_else(|| panic!("error : {} => {} ", "button", "nextSibling"));
        let _el11 = _el10
            .first_child()
            .unwrap_or_else(|| panic!("error : {} => {} ", "{block}", "firstChild"));
        ::leptos::leptos_dom::add_event_helper(
            ::leptos::wasm_bindgen::JsCast::unchecked_ref(&_el2),
            ::leptos::leptos_dom::ev::click,
            move |_| set_value(0),
        );
        ::leptos::leptos_dom::add_event_helper(
            ::leptos::wasm_bindgen::JsCast::unchecked_ref(&_el4),
            ::leptos::leptos_dom::ev::click,
            move |_| set_value.update(|value| *value -= step),
        );
        ::leptos::leptos_dom::mount_child(
            ::leptos::leptos_dom::MountKind::Before(&_el8.clone()),
            &{ { value } }.into_view(),
        );
        ::leptos::leptos_dom::add_event_helper(
            ::leptos::wasm_bindgen::JsCast::unchecked_ref(&_el10),
            ::leptos::leptos_dom::ev::click,
            move |_| set_value.update(|value| *value += step),
        );
        ::leptos::leptos_dom::View::Element(leptos::leptos_dom::Element {
            #[cfg(debug_assertions)]
            name: "div".into(),
            element: ::leptos::wasm_bindgen::JsCast::unchecked_into(root),
            #[cfg(debug_assertions)]
            view_marker: None,
        })
    }
}

